Snowflake documentation ITEM-10609 (#1037)
fusedio/fused-docs: docs/guide/data-input-outputs/import-connection/snowflake.mdx

---
id: snowflake
title: Snowflake Integration
sidebar_label: Snowflake Integration
sidebar_position: 4
---

# Snowflake Integration

Query data, browse metadata, read/write stages, and write DataFrames to tables in Snowflake -- all using Fused-managed OAuth tokens. No passwords or credentials are stored client-side.

## Prerequisites

### 1. Install the Snowflake extras

> ```bash

Reviewer: When does a suer need to do this?

In Workbench this isn't needed, so let's not show this as the first step

I would add this as optional step inside a collapsible prerequisite for people wanting to set this up outside of Workbench

pip install fused[snowflake]
```

This installs `snowflake-connector-python` alongside the core `fused` package. If you are using hosted Fused, this is already installed for you.

> ### 2. Connect your Snowflake account in the Fused workbench

Reviewer: ```suggestion
### 2. Connect your Snowflake account to Fused
```

> Before using the Python API, a Snowflake admin must set up a security integration, role, and service account user, and then you must complete the OAuth flow in the Fused UI.

Reviewer: Could we add where this needs to be run?

I'm guessing this is in some form of Snowflake admin panel?

**Step 2 -- Retrieve the client ID and secret:**

```sql
SELECT SYSTEM$SHOW_OAUTH_CLIENT_SECRETS('FUSED_OAUTH');
```

This returns a JSON object. Copy the `OAUTH_CLIENT_ID`, `OAUTH_CLIENT_SECRET`, and `OAUTH_CLIENT_SECRET_2` values -- you'll paste them into the Fused UI in the last step.

**Step 3 -- Create a role with warehouse usage:**

```sql
CREATE ROLE FUSED_WH_ROLE;
GRANT USAGE ON WAREHOUSE COMPUTE_WH TO ROLE FUSED_WH_ROLE;
```

**Step 4 -- Create a dedicated service account user and grant it the role:**

```sql
CREATE USER fused_service_account
  TYPE = PERSON
  LOGIN_NAME = 'fused_service_account'
  DISPLAY_NAME = 'Fused Service Account'
  DEFAULT_ROLE = FUSED_WH_ROLE
  DEFAULT_WAREHOUSE = COMPUTE_WH
  MUST_CHANGE_PASSWORD = FALSE
  PASSWORD = 'xxxxx';

GRANT ROLE FUSED_WH_ROLE TO USER fused_service_account;
```

:::warning
Do **not** sign in with an admin account during the OAuth step. Snowflake blocks `ACCOUNTADMIN` and `SECURITYADMIN` for custom OAuth integrations, which is why a dedicated service account user is required.
:::

> **Step 5 -- Connect in the Fused workbench:**

Reviewer: ```suggestion
**Step 5 -- Connect in Workbench**
```

> 1. Enable the **Snowflake integration** feature flag in **Settings > Experimental Features**.

Reviewer: ```suggestion
1. Enable the **Snowflake integration** feature flag in **[Preferences](/pr/1037/workbench/preferences) > Experimental Features**.
```

Author: ?? I think this will be a broken link

Reviewer: Oh indeed, I didn't truncate enough, I meant to add `/workbench/preferences` but it's fine

> 2. Open **Settings > Integrations** and click **Connect Snowflake**.

Reviewer: Could we add a screenshot of this please?

3. Enter your account identifier (e.g. `ORGNAME-ACCTNAME`), the `OAUTH_CLIENT_ID`, both client secrets, and optionally a role.
4. A Snowflake sign-in popup will open. Sign in with the **service account** user you created in Step 4 (not an admin account) to complete the OAuth flow.

---

## Quick start

```python
import fused.api

# One-liner: run a SQL query and get a DataFrame
df = fused.api.snowflake_query("SELECT * FROM my_db.my_schema.my_table LIMIT 10")
```

Or use the connection object for multiple operations:

```python
conn = fused.api.snowflake_connect(
    warehouse="COMPUTE_WH",
    database="MY_DB",
    schema="PUBLIC",
)

df = conn.query("SELECT * FROM sales WHERE region = 'US'")
```

---

> ## API Reference

Reviewer: We should add this in the API Reference

https://docs.fused.io/python-sdk/api-reference/api

This page in docs should only contains the step by step, API Reference is in another section

| `schema`    | `str \| None`  | `None`  | Default schema context.              |

```python
conn = fused.api.snowflake_connect(
    role="ANALYST",
    warehouse="COMPUTE_WH",
    database="ANALYTICS",
    schema="PUBLIC",
)
```

#### `fused.api.snowflake_query(sql, **kwargs) -> pd.DataFrame`

Execute a SQL query and return a DataFrame. This is a shorthand for `SnowflakeConnection(**kwargs).query(sql)`.

```python
df = fused.api.snowflake_query(
    "SELECT date, revenue FROM sales WHERE year = 2025",
    warehouse="COMPUTE_WH",
    database="ANALYTICS",
)
```

---

### `SnowflakeConnection` class

`fused.api.SnowflakeConnection` manages Fused-issued OAuth tokens and opens direct connections to Snowflake via `snowflake-connector-python`.

Each method call fetches a fresh short-lived token from the Fused server (cached for up to 10 minutes), opens a connection, executes the operation, and closes the connection.

#### Constructor

```python
from fused.api import SnowflakeConnection

conn = SnowflakeConnection(
    role="ANALYST",          # optional
    warehouse="COMPUTE_WH",  # optional
    database="MY_DB",        # optional
    schema="PUBLIC",          # optional
)

> ```

Reviewer 2: * This class is now FusedSnowflakeConnection
* Note that fused.api.snowflake_connect returns a regular SnowflakeConnection, not the Fused wrapped class described here

Write a pandas DataFrame to a Snowflake table using `snowflake.connector.pandas_tools.write_pandas`.

| Parameter | Type           | Default    | Description                                          |
|-----------|----------------|------------|------------------------------------------------------|
| `df`      | `pd.DataFrame` | (required) | The DataFrame to write.                              |
| `table`   | `str`          | (required) | Table name. Can be fully-qualified (`db.schema.tbl`).|
| `mode`    | `str`          | `"append"` | `"append"` or `"overwrite"`.                         |
| `**kwargs`|                |            | Extra args forwarded to `write_pandas`.              |

```python
import pandas as pd

df = pd.DataFrame({"id": [1, 2, 3], "value": [10.5, 20.3, 30.1]})

# Append rows to an existing table
conn.write(df, "ANALYTICS.PUBLIC.METRICS")

# Overwrite the table contents
conn.write(df, "ANALYTICS.PUBLIC.METRICS", mode="overwrite")

# Use a simple table name (uses database/schema from the connection)
conn = fused.api.snowflake_connect(database="ANALYTICS", schema="PUBLIC")
conn.write(df, "METRICS")
```

---

## Full example

> ```python

Reviewer: Could we make this a full UDF please?

I.e. something like

```
@fused.udf
def udf():
    import pandas as pd
    ...
```
